File: scraper.R
# Install necessary packages if not already installed
# install.packages(c("RSelenium", "rvest", "tidyverse"))

library(RSelenium)
library(rvest)
library(tidyverse)

# --- Configuration ---
TARGET_URL <- "https://example.com/cards" # REPLACE THIS with your actual target URL
CARD_SELECTOR <- ".card-item"             # REPLACE THIS with the actual CSS selector for a card container
SEE_MORE_BUTTON_SELECTOR <- "#see-more"   # REPLACE THIS with the actual 'See more results' button selector
# ---------------------

scrape_cards <- function() {
  
  # 1. Ask User for Input
  cat("Enter the maximum number of cards to scrape: ")
  max_cards <- as.integer(readLines("stdin", n = 1))
  
  if (is.na(max_cards) || max_cards <= 0) {
    stop("Invalid input. Please enter a positive integer.")
  }
  
  cat(sprintf("Targeting %d cards...\n", max_cards))
  
  # 2. Setup Selenium Driver
  # Assumes a Selenium server is running (e.g., via Docker) or use rsDriver for local
  # docker run -d -p 4445:4444 selenium/standalone-chrome
  tryCatch({
    remDr <- remoteDriver(remoteServerAddr = "localhost", port = 4545L, browserName = "chrome")
    remDr$open()
  }, error = function(e) {
    stop("Could not connect to Selenium server. Make sure it is running. Error: ", e$message)
  })
  
  on.exit(remDr$close()) # Ensure browser closes on exit
  
  # 3. Navigate to Page
  remDr$navigate(TARGET_URL)
  Sys.sleep(3) # Wait for initial load
  
  # 4. Check Initial Count
  get_card_count <- function() {
    elements <- remDr$findElements(using = "css selector", CARD_SELECTOR)
    return(length(elements))
  }
  
  current_count <- get_card_count()
  cat(sprintf("Initial load: %d cards found.\n", current_count))
  
  # 5. Conditional Logic
  if (current_count >= max_cards) {
    cat("Target reached with initial load. Scraping now...\n")
  } else {
    # Need more cards
    
    # Click "See more results" ONCE
    tryCatch({
      see_more_btn <- remDr$findElement(using = "css selector", SEE_MORE_BUTTON_SELECTOR)
      if (!is.null(see_more_btn)) {
        cat("Clicking 'See more results'...\n")
        see_more_btn$clickElement()
        Sys.sleep(2) # Wait for click action to process
      }
    }, error = function(e) {
      cat("Note: 'See more results' button not found or not clickable. Proceeding to scroll.\n")
    })
    
    # Infinite Scroll Loop
    last_count <- current_count
    no_change_counter <- 0
    MAX_NO_CHANGE_RETRIES <- 3
    
    cat("Entering infinite scroll mode...\n")
    
    while (current_count < max_cards) {
      # Scroll to bottom
      remDr$executeScript("window.scrollTo(0, document.body.scrollHeight);")
      Sys.sleep(2) # Wait for content to load (Adjust depending on connection speed)
      
      current_count <- get_card_count()
      cat(sprintf("Current count: %d\n", current_count))
      
      if (current_count == last_count) {
        no_change_counter <- no_change_counter + 1
        cat(sprintf("No new cards loaded (%d/%d)...\n", no_change_counter, MAX_NO_CHANGE_RETRIES))
        
        if (no_change_counter >= MAX_NO_CHANGE_RETRIES) {
          cat("Stopping: No more content could be loaded.\n")
          break
        }
      } else {
        no_change_counter <- 0 # Reset counter if new cards found
        last_count <- current_count
      }
    }
  }
  
  # 6. Extraction
  cat("Extracting data...\n")
  page_source <- remDr$getPageSource()[[1]]
  page_html <- read_html(page_source)
  
  cards <- page_html %>% html_nodes(CARD_SELECTOR)
  
  # Limit to requested number if we over-fetched
  if (length(cards) > max_cards) {
    cards <- cards[1:max_cards]
  }
  
  # --- Define Extractors Here ---
  # Replace these with actual selectors inside a card
  results <- map_dfr(cards, function(card) {
    title <- card %>% html_node(".card-title") %>% html_text(trim = TRUE)
    price <- card %>% html_node(".card-price") %>% html_text(trim = TRUE)
    
    tibble(
      Title = ifelse(is.na(title), "N/A", title),
      Price = ifelse(is.na(price), "N/A", price)
    )
  })
  
  cat(sprintf("Scraping complete. Collected %d items.\n", nrow(results)))
  return(results)
}

# Run the scraper
# df <- scrape_cards()
# print(df)

File: scraper.js
const puppeteer = require('puppeteer');
const readline = require('readline');

// --- CONFIGURATION ---
const TARGET_URL = 'https://www.autotrader.com/cars-for-sale/all-cars'; // Replace with actual URL
const CARD_SELECTOR = '[data-cmp="inventoryListing"]'; // Replace with actual card selector
const SEE_MORE_BUTTON_SELECTOR = 'button[data-cmp="loadMore"]'; // Replace with actual button selector
// ---------------------

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const askQuestion = (query) => {
    return new Promise(resolve => rl.question(query, resolve));
};

async function scrape() {
    console.log('--- Puppeteer Scraper ---');

    // 1. Ask User for Input
    const answer = await askQuestion('Enter the maximum number of cards to scrape: ');
    const maxCards = parseInt(answer, 10);

    if (isNaN(maxCards) || maxCards <= 0) {
        console.error('Invalid input. Please enter a positive integer.');
        rl.close();
        return;
    }

    console.log(`Targeting ${maxCards} cards...`);

    const browser = await puppeteer.launch({
        headless: false, // Set to true for headless mode
        defaultViewport: null,
        args: ['--start-maximized']
    });

    const page = await browser.newPage();

    try {
        // 2. Navigate to Page
        console.log(`Navigating to ${TARGET_URL}...`);
        await page.goto(TARGET_URL, { waitUntil: 'networkidle2', timeout: 60000 });

        // 3. Initial Count Check
        let currentCount = await page.$$eval(CARD_SELECTOR, els => els.length);
        console.log(`Initial visible cards: ${currentCount}`);

        if (currentCount >= maxCards) {
            console.log('Target reached with initial load. Scraping now...');
        } else {
            // 4. "See More" Logic
            const seeMoreBtn = await page.$(SEE_MORE_BUTTON_SELECTOR);
            if (seeMoreBtn) {
                console.log('Clicking "See more results"...');
                await seeMoreBtn.click();
                // Wait for some content change or timeout
                await new Promise(r => setTimeout(r, 3000));
            } else {
                console.log('"See more results" button not found. Proceeding to scroll mode.');
            }

            // 5. Infinite Scroll Loop
            console.log('Entering infinite scroll mode...');
            let previousCount = currentCount;
            let noChangeRetries = 0;
            const MAX_RETRIES = 3;

            while (currentCount < maxCards) {
                // Scroll to bottom
                await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));

                // Wait for content load (randomized delay to be safer)
                await new Promise(r => setTimeout(r, 2000 + Math.random() * 1000));

                // Check count
                currentCount = await page.$$eval(CARD_SELECTOR, els => els.length);
                console.log(`Loaded ${currentCount} cards...`);

                if (currentCount === previousCount) {
                    noChangeRetries++;
                    console.log(`No new items loaded (${noChangeRetries}/${MAX_RETRIES})...`);

                    if (noChangeRetries >= MAX_RETRIES) {
                        console.log('Stopping: No more new cards are loading.');
                        break;
                    }
                } else {
                    noChangeRetries = 0; // Reset retry counter
                    previousCount = currentCount;
                }
            }
        }

        // 6. Extraction
        console.log(`Scraping up to ${maxCards} items...`);
        const data = await page.$$eval(CARD_SELECTOR, (cards, limit) => {
            return cards.slice(0, limit).map(card => {
                // Modify these selectors based on the actual HTML structure of a card
                const titleEl = card.querySelector('h2') || card.querySelector('.title');
                const priceEl = card.querySelector('.price') || card.querySelector('[data-cmp="pricing"]');

                return {
                    title: titleEl ? titleEl.innerText.trim() : 'N/A',
                    price: priceEl ? priceEl.innerText.trim() : 'N/A',
                    // Add more fields here
                };
            });
        }, maxCards);

        console.log('--- Results ---');
        console.log(JSON.stringify(data, null, 2));
        console.log(`Successfully scraped ${data.length} items.`);

    } catch (error) {
        console.error('An error occurred during scraping:', error);
    } finally {
        await browser.close();
        rl.close();
    }
}

scrape();
